rx_signal_processing/rx_dsp_chain.cu

```diff
@@ -238,129 +218,124 @@ int main(int argc, char **argv){
     );
 
 
+
+    auto complex_taps = filters.get_filter_taps();
+
     DSPCore *dp = new DSPCore(&dsp_to_brian_begin, &dsp_to_brian_end, &dsp_to_data_write,
                              sig_options, sp_packet.sequence_num(), rx_rate, output_sample_rate,
-                             rx_freqs, &filters, beam_phases,
+                             rx_freqs, complex_taps, beam_phases,
                              beam_direction_counts, rx_metadata.initialization_time(),
-                             rx_metadata.sequence_start_time(), slice_ids);
+                             rx_metadata.sequence_start_time(), slice_ids, dm_rates);
 
     if (rx_metadata.numberofreceivesamples() == 0){
       //TODO(keith): handle error for missing number of samples.
     }
 
-    //We need to sample early to account for propagating samples through filters. The number of 
+
+    //We need to sample early to account for propagating samples through filters. The number of
     //required early samples is equal to the largest filter length in time (based on the rate at
-    //that stage.) The following lines find the max filter length in time and convert that to 
+    //that stage.) The following lines find the max filter length in time and convert that to
     //number of samples at the input rate.
-    int64_t extra_samples = (first_stage_dm_rate * second_stage_dm_rate * third_stage_dm_rate *
-                        filters.get_num_fourth_stage_taps());
 
-    if (extra_samples < (first_stage_dm_rate * second_stage_dm_rate * filters.get_num_third_stage_taps())) {
-      extra_samples = first_stage_dm_rate * second_stage_dm_rate * filters.get_num_third_stage_taps();
-    }
+    int64_t extra_samples = std::accumulate(dm_rates.begin(), dm_rates.end()-1, 1,
+                                            std::multiplies<int64_t>()) *
```

Comment:
Can we double check this is working as intended? Would it not multiply by the last filter order every time with line 242 being part of the operator (ie. first_stage_rate * last_num_taps * second_stage_rate * last_num_taps * third_stage_rate * last .... )

Reply:
I might suggest calculating total_dm_rate first, then dividing that by the last stage dm_rate and multiplying by last filter number of taps to get extra_samples value.

Reply:
I don't think so. I think accumulate tallies a result and then multiplies that result with the number of last stage taps. Accumulate works like pythons reduce.